pyro/distributions/nanmasked.py

```diff
@@ -0,0 +1,99 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import torch
+
+from .torch import MultivariateNormal, Normal
+
+
+class NanMaskedNormal(Normal):
+    """
+    Wrapper around :class:`~pyro.distributions.Normal` to allow partially
+    observed data as specified by NAN elements in :meth:`log_prob`; the
+    ``log_prob`` of these elements will be zero. This is useful for likelihoods
+    with missing data.
+
+    Example::
+
+        from math import nan
+        data = torch.tensor([0.5, 0.1, nan, 0.9])
+        with pyro.plate("data", len(data)):
+            pyro.sample("obs", NanMaskedNormal(0, 1), obs=data)
+    """
+
+    def log_prob(self, value: torch.Tensor) -> torch.Tensor:
+        ok = value.isfinite()
+        if ok.all():
+            return super().log_prob(value)
+
+        # Broadcast all tensors.
+        value, ok, loc, scale = torch.broadcast_tensors(value, ok, self.loc, self.scale)
+        result = value.new_zeros(value.shape)
+
+        # Evaluate ok elements.
+        if ok.any():
+            marginal = Normal(loc[ok], scale[ok], validate_args=False)
+            result[ok] = marginal.log_prob(value[ok])
+        return result
+
+
+class NanMaskedMultivariateNormal(MultivariateNormal):
+    """
+    Wrapper around :class:`~pyro.distributions.MultivariateNormal` to allow
+    partially observed data as specified by NAN elements in the argument to
+    :meth:`log_prob`. The ``log_prob`` of these events will marginalize over
+    the NAN elements. This is useful for likelihoods with missing data.
+
+    Example::
+
+        from math import nan
+        data = torch.tensor([
+            [0.1, 0.2, 3.4],
+            [0.5, 0.1, nan],
+            [0.6, nan, nan],
+            [nan, 0.5, nan],
+            [nan, nan, nan],
```

Comment:
I'm curious on what log prob of this all-nan element is?

Reply:
it will be zero since all values will be marginalized out.

Reply:
here, I just pushed a test for this fact 🙂